runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/InProcessEnvironmentFactory.java

```diff
@@ -0,0 +1,97 @@
+import java.util.concurrent.Future;
+import org.apache.beam.fn.harness.FnHarness;
+import org.apache.beam.model.pipeline.v1.Endpoints.ApiServiceDescriptor;
+import org.apache.beam.model.pipeline.v1.RunnerApi.Environment;
+import org.apache.beam.runners.fnexecution.GrpcFnServer;
+import org.apache.beam.runners.fnexecution.control.ControlClientPool;
+import org.apache.beam.runners.fnexecution.control.FnApiControlClientPoolService;
+import org.apache.beam.runners.fnexecution.control.InstructionRequestHandler;
+import org.apache.beam.runners.fnexecution.environment.EnvironmentFactory;
+import org.apache.beam.runners.fnexecution.environment.RemoteEnvironment;
+import org.apache.beam.runners.fnexecution.logging.GrpcLoggingService;
+import org.apache.beam.sdk.fn.stream.StreamObserverFactory;
+import org.apache.beam.sdk.fn.test.InProcessManagedChannelFactory;
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+
+/**
+ * An {@link EnvironmentFactory} that communicates to a {@link FnHarness} via the in-process gRPC
+ * channel.
+ *
+ * <p>TODO: Move this class to the runners/java-fn-execution module, with the Java SDK Harness as a
+ * provided dependency.
+ */
+class InProcessEnvironmentFactory implements EnvironmentFactory {
+
+  private final GrpcFnServer<GrpcLoggingService> loggingServer;
+  private final GrpcFnServer<FnApiControlClientPoolService> controlServer;
+
+  private final ControlClientPool.Source clientSource;
+
+  InProcessEnvironmentFactory(
+      GrpcFnServer<GrpcLoggingService> loggingServer,
+      GrpcFnServer<FnApiControlClientPoolService> controlServer,
+      ControlClientPool.Source clientSource) {
+    this.loggingServer = loggingServer;
+    this.controlServer = controlServer;
+    checkArgument(
+        loggingServer.getApiServiceDescriptor() != null,
+        "Logging Server cannot have a null %s",
+        ApiServiceDescriptor.class.getSimpleName());
+    checkArgument(
+        controlServer.getApiServiceDescriptor() != null,
+        "Control Server cannot have a null %s",
+        ApiServiceDescriptor.class.getSimpleName());
+    this.clientSource = clientSource;
+  }
+
+  @Override
+  public RemoteEnvironment createEnvironment(Environment container) throws Exception {
+    ExecutorService executor = Executors.newSingleThreadExecutor();
+    Future<?> fnHarness =
+        executor.submit(
+            () ->
+                FnHarness.main(
+                    PipelineOptionsFactory.create(),
+                    loggingServer.getApiServiceDescriptor(),
+                    controlServer.getApiServiceDescriptor(),
+                    InProcessManagedChannelFactory.create(),
+                    StreamObserverFactory.direct()));
+    executor.submit(
+        () -> {
+          try {
+            fnHarness.get();
+          } catch (Throwable t) {
+            executor.shutdownNow();
+          }
+        });
+
+    // TODO: find some way to populate the actual ID in FnHarness.main()
```

Comment:
Instead of relying on the WORKER_ID environment variable in Python, we could make the client id as part of the ApiServiceDescriptor. This would make the client id chosen by whomever vends the client and removes looking up a property.

@herohde @angoenka What do you think?

Reply:
I'm not sure I understand the context for this question. This ID is passed as part of the container contract and used for all portable API calls from that container. I think it's desirable that it's a single ID for simplicity and debuggability. Python's use of an environment variable to pass it around is an implementation detail and not, for example, how the Go SDK does it.

Reply:
The point I'm trying to make is that the ApiServiceDescriptor is meant to describe who and how your connecting to an endpoint. I can foresee that in the future you would specify things like credentials inside this ApiServiceDescriptor. Since talking to the service requires a client id, I felt as though it would be useful to put the client id there.

This client id may have value when doing job preparation/artifact uploading/... since it is a way to identify who you are talking to.

Reply:
I see what you're saying. It makes sense, but I'm not in favor of providing the same value in two different ways.

Reply:
I would be in favor of dropping how we currently provide the client id in exchange for this approach unless your referring to something else when you say `the same value in two different ways`

Reply:
+1, we should enhance `ApiServiceDescriptor` to hold connection information. I am in favor of adding `headers` in `ApiServiceDescriptor` instead of having a `client_id`.